src/nvim/tui/tui.c

```diff
@@ -844,7 +844,16 @@ static void fix_terminfo(TUIData *data)
   }
 
   if (STARTS_WITH(term, "xterm") || STARTS_WITH(term, "rxvt")) {
-    unibi_set_if_empty(ut, unibi_cursor_normal, "\x1b[?25h");
+    const char *normal = unibi_get_str(ut, unibi_cursor_normal);
+    if (!normal) {
+      unibi_set_str(ut, unibi_cursor_normal, "\x1b[?25h");
+    } else if (STARTS_WITH(normal, "\x1b[?12l")) {
+      // terminfo typically includes DECRST 12 as part of setting up the normal
+      // cursor, which interferes with the user's control via
+      // NVIM_TUI_ENABLE_CURSOR_SHAPE.  When DECRST 12 is present, skip over
+      // it, but honor the rest of the TI setting.
```

Comment:
> But still, aren't we just asking for the next problem where the user has no clue what's in their terminfo

They don't have to know.  This is adjusting the value for them.  It's just a stop-gap until we're able to switch to tickit, but it at least improves the experience for users that expect `NVIM_TUI_CURSOR_ENABLE_SHAPE=2 nvim` to do what we say it does.